### PR TITLE
Fix _list_field_strings

### PR DESCRIPTION
**Pull request by jneeven**
Fixes the bug mentioned in https://github.com/larq/zookeeper/pull/206#issuecomment-737396937. Also fixes another bug where printing a component that contains a tensor `Field` would throw a TensorFlow error, because the comparison with the parent value returned a boolean tensor rather than a boolean. This led TF to believe we were running in graph mode.

It would be good to add more unit tests for having an empty field with `allow_missing`, and for using tensors in various places, as both of these issues could have been identified through unit tests (we're probably not really testing the printing)

### CODE REVIEW

File: zookeeper/core/component.py [hunk 1]
```diff
@@ -289,14 +289,20 @@ def _list_field_strings(instance, color: bool, single_line: bool) -> Iterator[st
...
         parent_instance = next(
             utils.generate_component_ancestors_with_field(instance, field_name), None
         )
         if parent_instance is not None:
-            is_inherited = base_getattr(parent_instance, field_name) == value  # type: ignore
+            parent_value = base_getattr(parent_instance, field_name)  # type: ignore
+            if is_tensor(parent_value):
+                is_inherited = parent_value is value
+            else:
+                is_inherited = parent_value == value
```
Comment (jneeven) [inline zookeeper/core/component.py:305]: @AdamHillier  I'm not sure if we even need to make this distinction; if it really is an inherited value I think we could get away with `parent_value is value` in any case?

File: zookeeper/core/component.py [hunk 2]
```diff
@@ -289,14 +291,20 @@ def _list_field_strings(instance, color: bool, single_line: bool) -> Iterator[st
...
         else:
```
Comment (jneeven) [inline zookeeper/core/component.py:308]: @AdamHillier  do we even need this distinction? If it really is inherited, we could always just check `parent_value is value`, right?

File: zookeeper/core/component.py [hunk 3]
```diff
@@ -289,14 +291,20 @@ def _list_field_strings(instance, color: bool, single_line: bool) -> Iterator[st
         except AttributeError as e:
             if field.allow_missing:
                 value = utils.missing
+                yield f"{field_name}={value}" if color else f"{field_name}={value}"
+                continue
```
Comment (AdamHillier) [inline zookeeper/core/component.py:295]: Could you explain what the bug is here?

File: zookeeper/core/component.py (hunk 2 above)
Comment (AdamHillier) [reply zookeeper/core/component.py:308]: Hmm I'm slightly unclear on what's going on. What's wrong with always using the `==` check?

File: zookeeper/core/component.py (hunk 3 above)
Comment (jneeven) [reply zookeeper/core/component.py:295]: A few lines further down, you call `base_getattr` on this missing value, which throws an error.

File: zookeeper/core/component.py (hunk 2 above)
Comment (jneeven) [reply zookeeper/core/component.py:308]: `tensor == tensor` returns a tensor rather than a boolean, which breaks the logic

Comment (AdamHillier) [reply zookeeper/core/component.py:308]: Ah I see... That's super dumb, but makes sense, thanks :+1: I think I'd be okay with just using `parent_value is value`, actually, as you suggest.

File: zookeeper/core/component.py (hunk 3 above)
Comment (AdamHillier) [reply zookeeper/core/component.py:295]: Ah of course, nice find :+1:

Do you mind if we instead change `if parent_instance is not None:` to `if value is not utils.missing and parent_instance is not None:`?

Comment (jneeven) [reply zookeeper/core/component.py:295]: Yeah I somewhat do, if the value is `utils.missing` anyway, why would we even care to check what the parent instance is? It can't be inherited (or maybe it can, but it's meaningless). On the other hand it may make the code more maintainable if we don't have to duplicate that yield value, so it might not be too bad

Comment (AdamHillier) [reply zookeeper/core/component.py:293]: Ah but that's exactly what my change would do - skip the inheritence/parent instance check if it's a missing value. It'll just mean that we don't have to repeat the `yield` statement which I think is cleaner.

File: zookeeper/core/component.py [hunk 4]
```diff
@@ -289,14 +289,16 @@ def _list_field_strings(instance, color: bool, single_line: bool) -> Iterator[st
...
             else:
                 raise e from None
 
         parent_instance = next(
             utils.generate_component_ancestors_with_field(instance, field_name), None
         )
         if parent_instance is not None:
-            is_inherited = base_getattr(parent_instance, field_name) == value  # type: ignore
+            is_inherited = base_getattr(parent_instance, field_name) is value  # type: ignore
```
Comment (lgeiger) [inline zookeeper/core/component.py:301]: Could this be problematic if value is a literal value like an integer?
E.g. `0 is 1` would raise a warning: `SyntaxWarning: "is" with a literal. Did you mean "=="?`

File: zookeeper/core/component.py (hunk 3 above)
Comment (jneeven) [reply zookeeper/core/component.py:293]: We'd still run the following code:
```python
parent_instance = next(
            utils.generate_component_ancestors_with_field(instance, field_name), None
        )

if utils.is_component_instance(value):
            if is_inherited:
                value = "<inherited component instance>"
            elif single_line:
                value = repr(value)
            else:
                value = f"\n{INDENT}".join(str(value).split("\n"))
        elif is_inherited:
            value = "<inherited value>"
        elif callable(value):
            value = "<callable>"
        elif isinstance(value, str):
            value = f'"{value}"'
```
All of which is unnecessary. But sure, I'll do that anyway

File: zookeeper/core/component.py (hunk 4 above)
Comment (jneeven) [reply zookeeper/core/component.py]: Fixed in 7460528

Comment (AdamHillier) [reply zookeeper/core/component.py]: This is actually not necessary, the warning is printed only if one of the elements is a literal, not a variable. `a is b` is fine, even if `a` and `b` are both ints/floats/whatever.

image

So can we avoid adding the warning suppression?

File: zookeeper/core/component.py [hunk 5]
```diff
@@ -79,6 +79,7 @@ class C:
 
 import functools
 import inspect
+import warnings
 from typing import Any, Dict, Iterator, List, Optional, Type
```
Comment (jneeven) [inline zookeeper/core/component.py:83]:
```suggestion
from typing import Any, Dict, Iterator, List, Optional, Type
```

File: zookeeper/core/component.py [hunk 6]
```diff
@@ -295,8 +296,11 @@ def _list_field_strings(instance, color: bool, single_line: bool) -> Iterator[st
         parent_instance = next(
             utils.generate_component_ancestors_with_field(instance, field_name), None
         )
-        if parent_instance is not None:
-            is_inherited = base_getattr(parent_instance, field_name) == value  # type: ignore
+        if value is not utils.missing and parent_instance is not None:
+            # Suppress warning for comparing literals, e.g. `5 is 6`.
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+                is_inherited = base_getattr(parent_instance, field_name) is value  # type: ignore
```
Comment (jneeven) [inline zookeeper/core/component.py:303]:
```suggestion
            is_inherited = base_getattr(parent_instance, field_name) is value  # type: ignore
```

File: zookeeper/core/component.py [hunk 7]
```diff
@@ -295,8 +296,11 @@ def _list_field_strings(instance, color: bool, single_line: bool) -> Iterator[st
         parent_instance = next(
             utils.generate_component_ancestors_with_field(instance, field_name), None
         )
-        if parent_instance is not None:
-            is_inherited = base_getattr(parent_instance, field_name) == value  # type: ignore
+        if value is not utils.missing and parent_instance is not None:
+            # Suppress warning for comparing literals, e.g. `5 is 6`.
+            with warnings.catch_warnings():
```
Comment (jneeven) [inline zookeeper/core/component.py:301]:
```suggestion
        if value is not utils.missing and parent_instance is not None:
```

File: zookeeper/core/component.py [hunk 8]
```diff
@@ -295,8 +295,9 @@ def _list_field_strings(instance, color: bool, single_line: bool) -> Iterator[st
         parent_instance = next(
             utils.generate_component_ancestors_with_field(instance, field_name), None
         )
-        if parent_instance is not None:
-            is_inherited = base_getattr(parent_instance, field_name) == value  # type: ignore
+        if value is not utils.missing and parent_instance is not None:
+            # Suppress warning for comparing literals, e.g. `5 is 6`.
```
Comment (jneeven) [inline zookeeper/core/component.py:299]:
```suggestion
        if value is not utils.missing and parent_instance is not None:
```

**Review by AdamHillier — APPROVED** [review]
Thanks for the fix :)
